src/objects.c

```diff
@@ -51,24 +53,118 @@ P11PROV_OBJ *p11prov_obj_new(P11PROV_CTX *ctx, CK_SLOT_ID slotid,
     obj->slotid = slotid;
     obj->handle = handle;
     obj->class = class;
+    obj->cached = CK_INVALID_HANDLE;
 
     obj->refcnt = 1;
 
     return obj;
 }
 
-P11PROV_OBJ *p11prov_obj_ref(P11PROV_OBJ *obj)
+static void destroy_key_cache(P11PROV_OBJ *obj, P11PROV_SESSION *session)
+
+{
+    P11PROV_SESSION *_session = NULL;
+    CK_SESSION_HANDLE sess;
+    CK_RV ret;
+
+    if (obj->cached == CK_INVALID_HANDLE) {
+        return;
+    }
+
+    if (session) {
+        sess = p11prov_session_handle(session);
+    } else {
+        ret = p11prov_take_login_session(obj->ctx, obj->slotid, &_session);
+        if (ret != CKR_OK) {
+            P11PROV_debug("Failed to get login session. Error %lx", ret);
+            return;
+        }
+        sess = p11prov_session_handle(_session);
+    }
+
+    ret = p11prov_DestroyObject(obj->ctx, sess, obj->cached);
+    if (ret != CKR_OK) {
+        P11PROV_debug("Failed to destroy cached key. Error %lx", ret);
+    }
+    obj->cached = CK_INVALID_HANDLE;
+
+    if (_session) {
+        p11prov_return_session(_session);
+    }
+}
+
+static void cache_key(P11PROV_OBJ *obj)
+{
+    P11PROV_SESSION *session = NULL;
+    CK_BBOOL val_false = CK_FALSE;
+    CK_ATTRIBUTE template[] = { { CKA_TOKEN, &val_false, sizeof(val_false) } };
+    CK_SESSION_HANDLE sess;
+    CK_RV ret;
+
+    /* We cache only keys on the token */
+    if ((obj->class != CKO_PRIVATE_KEY && obj->class != CKO_PUBLIC_KEY)
+        || obj->cka_token != CK_TRUE) {
+        return;
+    }
+
+    ret = p11prov_take_login_session(obj->ctx, obj->slotid, &session);
+    if (ret != CKR_OK) {
+        P11PROV_debug("Failed to get login session. Error %lx", ret);
+        return;
+    }
+
+    /* If already cached, release and re-cache */
+    destroy_key_cache(obj, session);
+
+    sess = p11prov_session_handle(session);
+    ret = p11prov_CopyObject(obj->ctx, sess, obj->handle, template, 1,
+                             &obj->cached);
```

Comment:
I think the usage of the CopyObject and DestroyObject for caching should not be used by default. It might work ok with the software modules, but other modules are probably going to choke on that.

Reply:
Agree. Most HSM drivers do not support session object (except perhaps derived keys). This will fail on any key where the COPYABLE flag is false, or likely fails if e.g. if SENSITIVE is true. In general it is not even required that a module supports Copy object operation. If this kind of caching is enabled, it should be a quirk enabled only for white listed modules. Or better, filed as a performance bug against the module.

Reply:
I suppose the caching is optional as the error is mostly ignored with a debug level message. But attempting the copy always, may actually hit the device (behind a network) where it fails.
As minimum, failure to copy key should be cached and not attempted again.

Reply:
Yeah I can see adding a signal that caching doesn't work.
I wonder if this is something we can reliably detect per slot, so that we try at most once for devices that can't handle it.

Reply:
Should I check and trust CKA_COPYABLE ?
And default to false if it is not returned ?
Seem like a good indication.

Reply:
> I wonder if this is something we can reliably detect per slot, so that we try at most once for devices that can't handle it.

It will probably depend on the error. If you get UNSUPPORTED, its probably whole slot/module.  Other errors such as `CKR_ACTION_PROHIBITED` might mean this only for a object, but I never worked with this in pkcs11.

Reply:
I added a patch, let me know what you think